### PR TITLE
Beauty is now an element. Fixing an issue with enter/exit area comsigs.

### LINKED ISSUE #56978: Room Beauty doesn't improve
## Reproduction:
Go to meta robotics
check beauty 
image
try to remove everything in robotics, accidentally trail some oil footprints in
image
Clean all of the messy decals
image

Be a sad snob

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Converted Beauty into an element. It works fine, though I'm personally a bit nonplussed about that little hacky `proc/Detach` override I had to write down to make sure nothing bad happens if the element is attached to the same target twice.
I also needed to refactor code pertinent to COMSIG_EXIT_AREA and COMSIG_ENTER_AREA plus some more to solve a small problem with them not being called on movables with beauty when they effectively moved out of an area whilst bagged inside another movable. This should help with analogous edge cases by the by.
I think I have said enough. Maaaaybe I'll change how beauty element works ingame later, because shinies contributing to the overall area beauty score even if stashed away in an opaque container doesn't make sense, but I'm digressing.

## Why It's Good For The Game
See above. This will close #56978. I have tested this one. 

## Changelog
:cl:
fix: fixed a problem with room beauty not updating when movables are moved around whilst contained within another movable, or when deleted.
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
you're breathtaking

File: code/__DEFINES/dcs/signals.dm [hunk 1]
```diff
@@ -262,9 +262,9 @@
 
 /////////////////
 
-///from base of area/Entered(): (/area)
+///from base of area/Entered(): (/area). Sent to all contents that "area-sensitive", see __DEFINES/traits.dm for info.
 #define COMSIG_ENTER_AREA "enter_area"
-///from base of area/Exited(): (/area)
+///from base of area/Exited(): (/area). Idem.
```
Comment (Mothblocks) [inline code/__DEFINES/dcs/signals.dm:267]: What?

Comment (Mothblocks) [reply code/__DEFINES/dcs/signals.dm:267]: Okay looking it up it's a research term that looks like it means "ditto"...can you just copy and paste what you wrote above? 😅

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -339,6 +339,17 @@ Turf and target are separate in case you want to teleport some distance from a t
...
+//Returns a list of all locations (except the area) the movable is within.
+/proc/get_nested_locs(atom/movable/AM, include_turf = FALSE)
+	. = list()
+	var/atom/A = AM.loc
+	var/turf/T = get_turf(AM)
+	while(A && A != T)
+		. += A
+		A = A.loc
+	if(A && include_turf) //At this point, only the turf is left, if there is one.
+		. += A
```
Comment (Mothblocks) [inline code/__HELPERS/unsorted.dm:351]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/elements/beauty.dm [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	element_flags = ELEMENT_BESPOKE|ELEMENT_DETACH
+	id_arg_index = 2
+	var/beauty = 0
+	/**
+	  * Assoc list of atoms as keys and number of time the same element instance has been attached to them as assoc value.
+	  * So things don't get odd with same-valued yet dissimilar beauty modifiers being added to the same atom.
+	  */
+	var/beauty_counter = list()
+
+/datum/element/beauty/Attach(datum/target, _beauty)
```
Comment (Mothblocks) [inline code/datums/elements/beauty.dm:16]: Rename `_beauty` to `beauty`, and change `beauty = _beauty` to `src.beauty = beauty`. This lets us used named arguments.

File: code/datums/elements/beauty.dm [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	if(!beauty_counter[target] && ismovable(target))
+		var/atom/movable/mov_target = target
+		mov_target.become_area_sensitive(BEAUTY_ELEMENT_TRAIT)
+		RegisterSignal(mov_target, COMSIG_ENTER_AREA, .proc/enter_area)
+		RegisterSignal(mov_target, COMSIG_EXIT_AREA, .proc/exit_area)
+
+	beauty_counter[target]++
+
+	var/area/A = get_area(target)
```
Comment (Mothblocks) [inline code/datums/elements/beauty.dm:31]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/elements/beauty.dm [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	if(A && !A.outdoors)
+		A.totalbeauty += beauty
+		A.update_beauty()
+
+/datum/element/beauty/proc/enter_area(datum/source, area/A)
```
Comment (Mothblocks) [inline code/datums/elements/beauty.dm:36]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/elements/beauty.dm [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	SIGNAL_HANDLER
+
+	if(A.outdoors)
+		return
+	A.totalbeauty += beauty * beauty_counter[source]
+	A.update_beauty()
+
+/datum/element/beauty/proc/exit_area(datum/source, area/A)
```
Comment (Mothblocks) [inline code/datums/elements/beauty.dm:44]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/elements/beauty.dm [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	SIGNAL_HANDLER
+
+	if(A.outdoors)
+		return
+	A.totalbeauty -= beauty * beauty_counter[source]
+	A.update_beauty()
+
+/datum/element/beauty/Detach(datum/source, force)
+	var/area/A = get_area(source)
```
Comment (Mothblocks) [inline code/datums/elements/beauty.dm:53]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/area/areas.dm [hunk 8]
```diff
@@ -564,14 +564,16 @@ GLOBAL_LIST_EMPTY(teleportlocs)
...
+ * Sends signals COMSIG_AREA_ENTERED and COMSIG_ENTER_AREA (to a list of atoms)
  *
  * If the area has ambience, then it plays some ambience music to the ambience channel
  */
 /area/Entered(atom/movable/M)
 	set waitfor = FALSE
 	SEND_SIGNAL(src, COMSIG_AREA_ENTERED, M)
-	SEND_SIGNAL(M, COMSIG_ENTER_AREA, src) //The atom that enters the area
+	for(var/AM in M.area_sensitive_contents)
+		var/atom/movable/recipient = AM
```
Comment (Mothblocks) [inline code/game/area/areas.dm:575]:
```suggestion
	for(var/atom/movable/recipient as anything in M.area_sensitive_contents)
```

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#avoid-unnecessary-type-checks-and-obscuring-nulls-in-lists

File: code/game/area/areas.dm [hunk 9]
```diff
@@ -597,11 +599,13 @@ GLOBAL_LIST_EMPTY(teleportlocs)
...
  * Called when an atom exits an area
  *
- * Sends signals COMSIG_AREA_EXITED and COMSIG_EXIT_AREA (to the atom)
+ * Sends signals COMSIG_AREA_EXITED and COMSIG_EXIT_AREA (to a list of atoms)
  */
 /area/Exited(atom/movable/M)
 	SEND_SIGNAL(src, COMSIG_AREA_EXITED, M)
-	SEND_SIGNAL(M, COMSIG_EXIT_AREA, src) //The atom that exits the area
+	for(var/AM in M.area_sensitive_contents)
+		var/atom/movable/recipient = AM
```
Comment (Mothblocks) [inline code/game/area/areas.dm:607]:
```suggestion
	for(var/atom/movable/recipient as anything in M.area_sensitive_contents)
```

File: code/game/atoms_movable.dm [hunk 10]
```diff
@@ -535,6 +536,37 @@
 			return
 	A.Bumped(src)
 
+/atom/movable/Exited(atom/movable/AM, atom/newLoc)
+	. = ..()
+	if(AM.area_sensitive_contents)
+		for(var/k in get_nested_locs(src) + src)
+			var/atom/movable/location = k
```
Comment (Mothblocks) [inline code/game/atoms_movable.dm:543]:
```suggestion
		for(var/atom/movable/location as anything in get_nested_locs(src) + src)
```

File: code/game/atoms_movable.dm [hunk 11]
```diff
@@ -535,6 +536,37 @@
...
+			LAZYREMOVE(location.area_sensitive_contents, AM.area_sensitive_contents)
+
+/atom/movable/Entered(atom/movable/AM, atom/oldLoc)
+	. = ..()
+	if(AM.area_sensitive_contents)
+		for(var/k in get_nested_locs(src) + src)
+			var/atom/movable/location = k
```
Comment (Mothblocks) [inline code/game/atoms_movable.dm:550]:
```suggestion
		for(var/atom/movable/location as anything in get_nested_locs(src) + src)
```

File: code/game/atoms_movable.dm [hunk 12]
```diff
@@ -535,6 +536,37 @@
...
+			LAZYADD(location.area_sensitive_contents, AM.area_sensitive_contents)
+
+//See traits.dm. Use this in place of ADD_TRAIT.
+/atom/movable/proc/become_area_sensitive(trait_source = TRAIT_GENERIC)
+	if(!HAS_TRAIT(src, TRAIT_AREA_SENSITIVE))
+		RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_AREA_SENSITIVE), .proc/on_area_sensitive_trait_loss)
+		for(var/k in get_nested_locs(src) + src)
+			var/atom/movable/AM = k
```
Comment (Mothblocks) [inline code/game/atoms_movable.dm:558]: `as anything` and https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/atoms_movable.dm [hunk 13]
```diff
@@ -535,6 +536,37 @@
...
+			LAZYADD(AM.area_sensitive_contents, src)
+	ADD_TRAIT(src, TRAIT_AREA_SENSITIVE, trait_source)
+
+/atom/movable/proc/on_area_sensitive_trait_loss()
+	SIGNAL_HANDLER
+
+	UnregisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_AREA_SENSITIVE))
+	for(var/k in get_nested_locs(src) + src)
+		var/atom/movable/AM = k
```
Comment (Mothblocks) [inline code/game/atoms_movable.dm:567]: `as anything` and https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/atoms_movable.dm [hunk 14]
```diff
@@ -535,6 +536,37 @@
...
+			LAZYREMOVE(location.area_sensitive_contents, AM.area_sensitive_contents)
+
+/atom/movable/Entered(atom/movable/AM, atom/oldLoc)
+	. = ..()
+	if(AM.area_sensitive_contents)
+		for(var/k in get_nested_locs(src) + src)
+			var/atom/movable/location = k
+			LAZYADD(location.area_sensitive_contents, AM.area_sensitive_contents)
+
+//See traits.dm. Use this in place of ADD_TRAIT.
```
Comment (Mothblocks) [inline code/game/atoms_movable.dm:553]:
```suggestion
/// See traits.dm. Use this in place of ADD_TRAIT.
```

Comment (Mothblocks) [reply code/game/atoms_movable.dm:553]: The important part is the `///`, I just added an extra space because that's what I like 🙂

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Recheck yo contributing guidelines

File: code/__DEFINES/dcs/signals.dm (hunk 1 above)
Comment (Ghommie) [reply code/__DEFINES/dcs/signals.dm:267]: Whoops.

**Review by Mothblocks — APPROVED** [review]

**Pull request merged**